src/util/bip152.rs

```diff
@@ -0,0 +1,353 @@
+    pub prefilled_txs: Vec<PrefilledTransaction>,
+}
+impl_consensus_encoding!(HeaderAndShortIds, header, nonce, short_ids, prefilled_txs);
+
+impl HeaderAndShortIds {
+    /// Create a new HeaderAndShortIds from a full block.
+    ///
+    /// The version number must be either 1 or 2.
+    ///
+    /// The [prefill] slice indicates which transactions should be prefilled in
+    /// the block. It should contain the indexes in the block of the txs to
+    /// prefill. It must be in ordered. 0 should not be included as the
+    /// coinbase tx is always prefilled.
+    ///
+    /// > Nodes SHOULD NOT use the same nonce across multiple different blocks.
+    ///
+    /// Possible [Error] variants:
+    /// - UnknownVersion: If the version is not 1 or 2.
+    /// number of transactions in the block.
+    pub fn from_block(
+        block: &Block,
+        nonce: u64,
+        version: u32,
+        mut prefill: &[usize],
+    ) -> Result<HeaderAndShortIds, Error> {
+        if version != 1 && version != 2 {
+            return Err(Error::UnknownVersion);
+        }
+
+        let siphash_keys = ShortId::calculate_siphash_keys(&block.header, nonce);
+
+        let mut prefilled = vec![];
+        let mut short_ids = vec![];
+        let mut last_prefill = 0;
+        for (idx, tx) in block.txdata.iter().enumerate() {
+            // Check if we should prefill this tx.
+            let prefill_tx = if prefill.get(0) == Some(&idx) {
+                prefill = &prefill[1..];
+                true
+            } else {
+                idx == 0 // Always prefill coinbase.
+            };
+
+            if prefill_tx {
+                let diff_idx = idx - last_prefill;
+                last_prefill = idx;
+                prefilled.push(PrefilledTransaction(
+                    diff_idx as u64,
+                    match version {
+                        1 => {
+                            // strip witness for version 1
+                            let mut no_witness = tx.clone();
+                            no_witness.input.iter_mut().for_each(|i| i.witness.clear());
+                            no_witness
+                        }
+                        2 => tx.clone(),
+                        _ => unreachable!(),
+                    },
+                ));
+            } else {
+                short_ids.push(ShortId::with_siphash_keys(
+                    &match version {
+                        1 => tx.txid().as_hash(),
+                        2 => tx.wtxid().as_hash(),
+                        _ => unreachable!(),
+                    },
+                    siphash_keys,
+                ));
+            }
+        }
+
+        Ok(HeaderAndShortIds {
+            header: block.header.clone(),
+            nonce: nonce,
+            // Provide coinbase prefilled.
+            prefilled_txs: prefilled,
+            short_ids: short_ids,
+        })
+    }
+}
+
+/// A BlockTransactionsRequest structure is used to list transaction indexes
```

Comment:
Hmm, I usually do square brackets because that (supposedly) allows rustdocs to make autolinks.

Reply:
Ah, I don't have a problem with that either, links are always better. I just saw a few occurrences where there were none. While this could be seen as  valid if you reference the data structure from the BIP instead of our types it seems like a useless distinction.

Reply:
Yeah my Rust coding style changed a bit over the last year :sweat_smile: I changed them all to square brackets in the entire module.